src/main/scala/org/apache/spark/shuffle/ucx/NvkvWrapper.scala

```diff
@@ -0,0 +1,317 @@
+package org.apache.spark.shuffle.ucx
+
+import java.util.zip.CRC32
+import scala.collection.mutable.ListBuffer
+import org.openucx.jnvkv._
+import org.openucx.jucx.ucp._
+import org.openucx.jucx.UcxException
+import org.openucx.jucx.ucs.UcsConstants
+import java.nio.ByteBuffer
+import java.nio.ByteOrder
+import org.apache.spark.SparkEnv
+import org.apache.spark.shuffle.utils.{UnsafeUtils, CommonUtils}
+import org.apache.spark.internal.Logging
+import java.nio.BufferOverflowException
+import org.openucx.jnvkv.NvkvException
+
+object NvkvWrapper {
+  private var nvkv: NvkvWrapper = null
+
+  def getNvkv(ucxContext: UcpContext, numOfPartitions: Int): NvkvWrapper = {
+    if (nvkv == null) nvkv = new NvkvWrapper(ucxContext, numOfPartitions)
+    nvkv
+  }
+}
+
+class NvkvWrapper private(ucxContext: UcpContext, private var numOfPartitions: Long) extends Logging {
+  final private val nvkvWriteBufferSize  = 1024*1024*40L
+  final private val nvkvReadBufferSize  = 1024*1024*40L
+  final private val nvkvRemoteReadBufferSize = 1024*1024*43L
+  final private val nvkvNumOfReadBuffers = 16
+  final private val alignment = 512
+  private var nvkvWriteBuffer: ByteBuffer = null
+  private var nvkvReadBuffer: ByteBuffer = null
+  private var nvkvRemoteReadBuffer: ByteBuffer = null
+  private var nvkv: Nvkv.Context = null
+  private var ds_idx = 0
+  private var nvkvStorageSize = 0L
+  private var partitionSize = 0L
+  private var packData: ByteBuffer = null
+  val numOfMappers  = SparkEnv.get.conf.getInt("spark.groupByTest.numMappers", 1)
+  val numOfReducers = SparkEnv.get.conf.getInt("spark.groupByTest.numReducers", 1)
+  private var reducePartitions: Array[Array[ReducePartition]] = Array.ofDim[ReducePartition](numOfMappers, numOfReducers)
+  val nvkvLogEnabled = SparkEnv.get.conf.getInt("spark.nvkvLogs.enabled", 0)
+  val blockManager = SparkEnv.get.blockManager.blockManagerId
+  val executerId = blockManager.executorId.toInt
+  var core_mask = (1 << (executerId-1)).toHexString
+
+  private def nvkvLogDebug(msg: => String) {
+    if (nvkvLogEnabled == 1) {
+      logDebug(msg)
+    }
+  }
+
+  logDebug(s"LEO NvkvWrapper constructor cpu_mask ${(1 << (executerId-1)).toHexString}")
+  try {
+    Nvkv.init("mlx5_0", 0, (1 << (executerId-1)).toHexString) 
```

Comment:
Document

Reply:
Unresolved

Reply:
Check `core_mask` declaration, I've added doc above it.